crypto/rsa/rsa_lib.c

```diff
@@ -991,6 +991,10 @@ int EVP_PKEY_CTX_set_rsa_pss_keygen_md_name(EVP_PKEY_CTX *ctx,
  */
 int EVP_PKEY_CTX_set_rsa_oaep_md(EVP_PKEY_CTX *ctx, const EVP_MD *md)
 {
+    /* If key type not RSA return error */
+    if (!EVP_PKEY_CTX_is_a(ctx, "RSA"))
+        return -1;
+
```

Comment:
This feels a little bit misplaced? What happens if you do not check here? IMO the provider or the ctrl translation code should reject.

Reply:
If you look in this file you will see similiar checks..

Reply:
Yes, there are some, should they be there though?

Reply:
Without the check things pass..
RSA OEAP MD uses "digest" as the key...
So for example using this on an SM2 key will set the digest param, which is thinks is fine.. So it passes. So I think this rules out putting it in the provider..

Some of the tests in this file are checking RSA and RSA-PSS since they are seperate types. Moving this into the ctrl layer might be a bit hard.


evp_pkey_ctx_ctrl_int() calls evp_pkey_ctx_ctrl_to_param()
evp_pkey_ctx_ctrl_to_param() passes a pctx and a keytype (nid), but it doesnt check if they match.. I think this works in most cases because the param keys are unique for lots of cases. 

Are you suggesting converting the nid to a name inside evp_pkey_ctx_ctrl_to_param()? I am bit scared of the scope of that change..

Reply:
Tried doing this, but EVP_PKEY_CTX_set_ec_paramgen_curve_nid() fails when SM2 is used :(.. So this isnt going to work generally.

Reply:
I assume RSA and RSA_PSS will have a similiar problem.. We also have DH and DHX.